src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -325,6 +356,459 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
 
 
 /*
+ * AllRelationsJoinedOnPartitionKey expands the var equivalances gathered from
+ * standard planner's relation and join hooks. The main goal of this function is
+ * to deduce whether all base relations in the query are joined on their partition
+ * columns or not.
+ *
+ * The function short circuits if there exists a single base relation in the query
+ * and returns true.
+ *
+ * Also reference tables, which don't have partition columns, treated separately such
+ * that they are always counted as joined on the partition key.
+ */
+static bool
+AllRelationsJoinedOnPartitionKey(RelationRestrictionContext *restrictionContext,
+								 JoinRestrictionContext *joinRestrictionContext)
+{
+	List *innerJoinVarEquivalences = NIL;
+	List *specialJoinVarEquivalences = NIL;
+	List *varEquivalences = NIL;
+	VarEquivalenceClass *commonEquivalenceClass = NULL;
+	bool allRelationsJoinedOnPartitionKey = true;
+	uint32 referenceRelationCount = ReferenceRelationCount(restrictionContext);
+	uint32 totalRelationCount = list_length(restrictionContext->relationRestrictionList);
+
+	ListCell *commonEqClassCell = NULL;
+	ListCell *restrictionCell = NULL;
+	Relids commonRelids = NULL;
+
+	/*
+	 * If there is a single RTE_RELATION that is not a reference table,
```

Comment:
this means there are no joins between non-reference tables. Comment indirectly says that however it requires some thinking to understand.